sdk/webpubsub/azure-webpubsub-client/azure/webpubsub/client/_client.py

```diff
@@ -0,0 +1,864 @@
+            _LOGGER.debug("wait for seq thread stop")
+            old_thread_seq_ack.join()
+        if old_thread and old_thread.is_alive():
+            _LOGGER.debug("wait for listener thread stop")
+            old_thread.join()
+
+        _LOGGER.info("stop client successfully")
+
+    @overload
+    def on(
+        self, event: Union[Literal[CallBackType.CONNECTED], str], listener: Callable[[OnConnectedArgs], None]
+    ) -> None:
+        """Add handler for connected event.
+        :param event: The event name. Required.
+        :type event: str
+        :param listener: The handler
+        :type listener: callable.
+        """
+
+    @overload
+    def on(
+        self, event: Union[Literal[CallBackType.DISCONNECTED], str], listener: Callable[[OnDisconnectedArgs], None]
+    ) -> None:
+        """Add handler for disconnected event.
+        :param event: The event name. Required.
+        :type event: str
+        :param listener: The handler
+        :type listener: callable.
+        """
+
+    @overload
+    def on(self, event: Union[Literal[CallBackType.STOPPED], str], listener: Callable[[], None]) -> None:
+        """Add handler for stopped event.
+        :param event: The event name. Required.
+        :type event: str
+        :param listener: The handler
+        :type listener: callable.
+        """
+
+    @overload
+    def on(
+        self,
+        event: Union[Literal[CallBackType.SERVER_MESSAGE], str],
+        listener: Callable[[OnServerDataMessageArgs], None],
+    ) -> None:
+        """Add handler for server messages.
+        :param event: The event name. Required.
+        :type event: str
+        :param listener: The handler
+        :type listener: callable.
+        """
+
+    @overload
+    def on(
+        self, event: Union[Literal[CallBackType.GROUP_MESSAGE], str], listener: Callable[[OnGroupDataMessageArgs], None]
+    ) -> None:
+        """Add handler for group messages.
+        :param event: The event name. Required.
+        :type event: str
+        :param listener: The handler
+        :type listener: callable.
+        """
+
+    @overload
+    def on(
+        self,
+        event: Union[Literal[CallBackType.REJOIN_GROUP_FAILED], str],
+        listener: Callable[[OnRejoinGroupFailedArgs], None],
+    ) -> None:
+        """Add handler for rejoining group failed.
+        :param event: The event name. Required.
+        :type event: str
+        :param listener: The handler
+        :type listener: callable.
+        """
+
+    @distributed_trace
+    def on(
+        self, event: Union[CallBackType, str], listener: Callable, **kwargs: Any  # pylint: disable=unused-argument
+    ) -> None:
+        if event in self._handler:
+            self._handler[event].append(listener)
+        else:
+            _LOGGER.error("wrong event type: %s", event)
+
+    @overload
+    def off(
+        self, event: Union[Literal[CallBackType.CONNECTED], str], listener: Callable[[OnConnectedArgs], None]
```

Comment:
These `off` methods are going to be hard to use - we should investigate a better pattern here.
For example, unless I have the exact instance of the `Callable` that I passed into the `on` method - this wont work.

Reply:
If the target listener doesn't exist, we will log info; If it exists, remove it; If event type is not right, we log error but don't raise exception. No matter the listener exists or not, we don't raise exception. It keeps consistency across language. Is there any ambiguity?

Reply:
The error behaviour sounds fine, but what I meant is that if I have code like in the readme:
```python
client.on("group-message", lambda e: print(f"Received message: {e.data}"))
```
There is no way to turn this handler "off". I have to actually hold onto an instance of the lamda or callable in order to remove it.
It's fine for first preview - but we have to revisit this.